### PR TITLE
[got] introduce typings

### PR DESCRIPTION
**Pull request by BendingBender**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

File: types/got/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+declare function got(url: got.GotUrl, options: got.GotJSONOptions & http.RequestOptions): got.GotPromise<object>;
+declare function got(url: got.GotUrl, options: got.GotFormOptions<string> & http.RequestOptions): got.GotPromise<string>;
+declare function got(url: got.GotUrl, options: got.GotFormOptions<null> & http.RequestOptions): got.GotPromise<Buffer>;
+declare function got(url: got.GotUrl, options: got.GotBodyOptions<string> & http.RequestOptions): got.GotPromise<string>;
+declare function got(url: got.GotUrl, options: got.GotBodyOptions<null> & http.RequestOptions): got.GotPromise<Buffer>;
+// tslint:enable unified-signatures
+
+declare namespace got {
+    // tslint:disable unified-signatures
+    function get(url: GotUrl): GotPromise<string>;
```
Comment (andy-ms) [inline types/got/index.d.ts]: You could define:
```ts
interface Fn { //name me better
    (url: GotUrl): GotPromise<String>;
    (url: GotUrl, options: GotJSONOptions & http.RequestOptions): GotPromise<object>;
    (url: GotUrl, options: GotFormOptions<string> & http.RequestOptions): GotPromise<string>;
    (url: GotUrl, options: GotFormOptions<null> & http.RequestOptions): GotPromise<Buffer>;
    (url: GotUrl, options: GotBodyOptions<string> & http.RequestOptions): GotPromise<string>;
    (url: GotUrl, options: GotBodyOptions<null> & http.RequestOptions): GotPromise<Buffer>;
}
const get: Fn;
const post: Fn;
...
```

File: types/got/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    // function delete(url: GotUrl): GotPromise<string>;
+    // function delete(url: GotUrl, options: GotJSONOptions & http.RequestOptions): GotPromise<object>;
+    // function delete(url: GotUrl, options: GotFormOptions<string> & http.RequestOptions): GotPromise<string>;
+    // function delete(url: GotUrl, options: GotFormOptions<null> & http.RequestOptions): GotPromise<Buffer>;
+    // function delete(url: GotUrl, options: GotBodyOptions<string> & http.RequestOptions): GotPromise<string>;
+    // function delete(url: GotUrl, options: GotBodyOptions<null> & http.RequestOptions): GotPromise<Buffer>;
+
+    // tslint:enable unified-signatures
+
+    const stream: stream;
```
Comment (andy-ms) [inline types/got/index.d.ts]:
```ts
type StreamFn = (url: string | http.RequestOptions | URL, options?: GotStreamFullOptions & http.RequestOptions) => GotEmitter & nodeStream.Duplex;
const stream: StreamFn & Record<"get" | "post" | "put" | "patch" | "head" | "delete", StreamFn>;
```

File: types/got/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        json: true;
+    }
+
+    interface GotFormOptions<E extends string | null> extends GotStreamOptions<E> {
+        body?: {[key: string]: any};
+        form: true;
+        json?: boolean;
+    }
+
+    // http.RequestOptions &
```
Comment (andy-ms) [inline types/got/index.d.ts]: What's preventing this?

Comment (BendingBender) [inline types/got/index.d.ts]: nothing :-)

Comment (andy-ms) [inline types/got/index.d.ts]: You should either do it or delete the comment, don't leave me hanging!

Comment (BendingBender) [inline types/got/index.d.ts]: did it, sorry for not being clear

Comment (andy-ms) [inline types/got/index.d.ts]: Thanks, can this comment be removed then?

Comment (BendingBender) [inline types/got/index.d.ts]: yes. change committed, too.

Comment (andy-ms) [inline types/got/index.d.ts]: I'm still seeing this comment in the latest version, did you forget to push?

Comment (BendingBender) [inline types/got/index.d.ts]: OK, I'm sorry, I misunderstood your comment. Fixed now and pushed.

File: types/got/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        (url: GotUrl, options: GotFormOptions<string>): GotPromise<string>;
+        (url: GotUrl, options: GotFormOptions<null>): GotPromise<Buffer>;
+        (url: GotUrl, options: GotBodyOptions<string>): GotPromise<string>;
+        (url: GotUrl, options: GotBodyOptions<null>): GotPromise<Buffer>;
+    }
+    // tslint:disable unified-signatures
+
+    type GotStreamFn = (url: string | http.RequestOptions | URL, options?: GotOptions<string | null>) => GotEmitter & nodeStream.Duplex;
+
+    type GotUrl = string | http.RequestOptions | URL;
```
Comment (andy-ms) [inline types/got/index.d.ts]: This could be used in the first parameter of `GotStreamFn`

**Comment by andy-ms** [comment]
Thanks!

**Comment by BendingBender** [comment]
Thank you for your review!

**Pull request merged**